perp/man/sissylog.8
.\" sissylog.8
.\" wcm, 2009.09.29 - 2011.02.01
.\" ===
.TH sissylog 8 "March 2011" "perp-2.04" "persistent process supervision"
.SH NAME
sissylog \- log stdin to
.BR syslog (3)
.SH SYNOPSIS
.B sissylog [\-hV] [
.I ident
.B [
.I facility
.B ]]
.SH DESCRIPTION
.B sissylog
reads lines from standard input and writes them to the system logger with
.BR syslog (3).
.PP
If an
.I ident
argument is specified,
.B sissylog
prepends log entries accordingly.
Otherwise, entries are prepended with the identity string
``sissylog''.
.PP
If a
.I facility
argument is specified,
it should be in the form of a string selected from the following list:
.BR LOG_AUTH ,
.BR LOG_CRON ,
.BR LOG_DAEMON ,
.BR LOG_FTP ,
.BR LOG_KERN ,
.BR LOG_LPR ,
.BR LOG_MAIL ,
.BR LOG_NEWS ,
.BR LOG_SYSLOG ,
.BR LOG_USER ,
or
.BR LOG_UUCP .
.B sissylog
will use a corresponding facility argument with
.BR openlog (8).
If the 
.I facility
argument is not specified or not recognized,
.B sissylog
will use the facility
.B LOG_DAEMON
by default.
.PP
.B sissylog
inspects the beginning of each line for
.BR alert: ,
.BR error: ,
.BR warning: ,
.BR notice: ,
or
.BR debug: ,
and sets the priority of each log entry accordingly.
Otherwise, entries are made with the
.B LOG_INFO
priority by default.
.PP
.B sissylog
splits lines longer than 800 characters,
prepending continued lines with a `+' character.
.B sissylog
does not log empty lines,
and converts unprintable control characters to `?'.
.SH OPTIONS
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.SH SEE ALSO
.nh
.BR openlog(3),
.BR syslog(3),
.BR perp_intro (8),
.BR perpboot (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpok (8),
.BR perpstat (8),
.BR tinylog (8)
.\" EOF sissylog.8


CHANGES
# CHANGES
# perp release history and changelog
# wcm, 2010.02.17 - 2011.03.22
# ===

perp-2.04 (2011.03.22):
 * perphup:
   - added option -t, trigger perpd shutdown
 * perpboot:
   - waitpid() for deux[1] on SIGTERM
 * perp-setup:
   - added arg for ctldir
   - configure symlink /etc/perp/.control --> ctldir
   - configure rc.perp bootscript for runchoom(8)
 * runchoom:
   - new, utility for linux "oom killer" abatement

perp-2.03 (2011.03.11):
 * perp-setup:
   - in rc.perp boot script for perpd(8), setup default runlimit(8)
     configuration to hard limits for RLIMIT_NPROC, RLIMIT_NOFILE

perp-2.02 (2011.02.22):
 * perpd:
   - eliminated some spurious/redundant client connection logging

perp-2.01 (2011.02.11):
 * perp-setup:
   - runlimit added to .boot/rc.perp script
 * perpd:
   - added option -g, enable group permission on control socket
   - additional comments in LIMITS section of perpd(8) manual
 * runlimit:
   - added option -E, take resource limits from environment
   - added option -F, take resource limits from file
 * lasagna:
   - outvec/mk_outvec.c: portability fix for dietlibc

perp-2.00 (2011.02.02):
 * perpd:
   - coalesce supervisor (formerly perpetrate(8)) into scanner/controller
   - runs as a single process
   - all ipc with client applications via single domain socket 
   - ipc protocol version 2 (see PROTO_V2.txt)
 * perpetrate:
   - eliminate perpetrate(8) executable (subsumed by perpd(8))
   - perpetrate(5) manual page retained for runscript conventions
 * perpctl:
   - add commands [A | X], activate/deactivate service
 * libasagna:
   - nuscan.h, improve numeric scanning interface (supercedes nscan.h)
   - other iterative improvements incorporated from projects hdb and ezcdb
 
perp-0.00 (2010.01.12):
 - maiden public release

### EOF


perp/sissylog.c
/* sissylog.c
** log stdin to syslog
** wcm, 2009.09.28 - 2009.12.06
** ===
*/
#include <stddef.h>

#include <unistd.h>
#include <syslog.h>

#include "cstr.h"
#include "ioq.h"
#include "ioq_std.h"
#include "nextopt.h"

#include "sissylog.h"

static const char *progname = NULL;
static const char prog_usage[] = " [-hV] [ ident [ facility ]]";

/* ioq for stdin: */
#define INBUF_SIZE  1024
uchar_t        inbuf[INBUF_SIZE];
static ssize_t read_op(int fd, void *buf, size_t len);
ioq_t          in = ioq_INIT(0, inbuf, sizeof inbuf, &read_op);

/* buffer for sissylog's own messages: */
static char    attention[200];

/* other declarations in scope: */
static void    logline_post(const char *logline, int flag_continue);
static int     do_log(void);


static         ssize_t read_op(int fd, void *buf, size_t len)
{
   ssize_t        r = 0;

   do {
      r = read(fd, buf, len);
   } while ((r == -1) && (errno == EINTR));

   return r;
}


static
void logline_post(const char *logline, int flag_continue)
{
   static int     priority;

   if (flag_continue) {
      syslog(priority, "+%s", logline);
   } else {
      priority = LOG_INFO;
      if (cstr_contains(logline, "alert:")) {
         priority = LOG_ALERT;
      } else if (cstr_contains(logline, "error:")) {
         priority = LOG_ERR;
      } else if (cstr_contains(logline, "warning:")) {
         priority = LOG_WARNING;
      } else if (cstr_contains(logline, "notice:")) {
         priority = LOG_NOTICE;
      } else if (cstr_contains(logline, "debug:")) {
         priority = LOG_DEBUG;
      }
      syslog(priority, "%s", logline);
   }

   return;
}


static
int do_log(void)
{
   char           logline[LOGLINE_MAX + 1];
   size_t         len = 0;
   ssize_t        r = 0;
   char           c;
   int            flag_continue = 0;

   for (;;) {
      r = ioq_GETC(&in, &c);
      if (r < 1) {
         /* eof or io error: */
         if (len > 0) {
            /* post partial line: */
            logline[len] = '\0';
            logline_post(logline, flag_continue);
            flag_continue = 0;
            len = 0;
         }
         break;
      }
      if (c == '\n') {
         logline[len] = '\0';
         if (len > 0) {
            logline_post(logline, flag_continue);
         }
         flag_continue = 0;
         len = 0;
         continue;
      }
      /* filter control characters other than \t: */
      if ((c < 32) && (c != '\t'))
         c = '?';
      logline[len++] = c;
      if (len == LOGLINE_MAX) {
         logline[len] = '\0';
         logline_post(logline, flag_continue);
         flag_continue = 1;
         len = 0;
         continue;
      }
   }

   return r;
}


int main(int argc, char *argv[])
{
   char           opt;
   nextopt_t      nopt = nextopt_INIT(argc, argv, ":hV");
   const char    *arg_ident;
   const char    *arg_facility = "LOG_DAEMON";
   int            id_facility;
   int            e;

   progname = nextopt_progname(&nopt);
   arg_ident = progname;
   while ((opt = nextopt(&nopt))) {
      char           optc[2] = { nopt.opt_got, '\0' };
      switch (opt) {
      case 'h':
         usage();
         die(0);
         break;
      case 'V':
         version();
         die(0);
         break;
      case '?':
         if (nopt.opt_got != '?') {
            fatal_usage("invalid option: -", optc);
         }
         /* else fallthrough: */
      default:
         die_usage();
         break;
      }
   }

   argc -= nopt.arg_ndx;
   argv += nopt.arg_ndx;

   if (argv[0]) {
      arg_ident = argv[0];
      if (argv[1]) {
         arg_facility = argv[1];
      }
   }

   if (cstr_cmp(arg_facility, "LOG_AUTH") == 0) {
      id_facility = LOG_AUTH;
   } else if (cstr_cmp(arg_facility, "LOG_CRON") == 0) {
      id_facility = LOG_CRON;
   } else if (cstr_cmp(arg_facility, "LOG_DAEMON") == 0) {
      id_facility = LOG_DAEMON;
   } else if (cstr_cmp(arg_facility, "LOG_FTP") == 0) {
      id_facility = LOG_FTP;
   } else if (cstr_cmp(arg_facility, "LOG_KERN") == 0) {
      id_facility = LOG_KERN;
   } else if (cstr_cmp(arg_facility, "LOG_LPR") == 0) {
      id_facility = LOG_LPR;
   } else if (cstr_cmp(arg_facility, "LOG_MAIL") == 0) {
      id_facility = LOG_MAIL;
   } else if (cstr_cmp(arg_facility, "LOG_NEWS") == 0) {
      id_facility = LOG_NEWS;
   } else if (cstr_cmp(arg_facility, "LOG_SYSLOG") == 0) {
      id_facility = LOG_SYSLOG;
   } else if (cstr_cmp(arg_facility, "LOG_USER") == 0) {
      id_facility = LOG_USER;
   } else if (cstr_cmp(arg_facility, "LOG_UUCP") == 0) {
      id_facility = LOG_UUCP;
   } else {
      id_facility = LOG_DAEMON;
   }

   openlog(arg_ident, 0, id_facility);
   cstr_vcopy(attention, progname, ": logging from stdin ...");
   logline_post(attention, 0);

   /* loop reading stdin: */
   e = do_log();

   /* done: eof or i/o error: */
   if (e) {
      cstr_vcopy(attention, "notice: ", progname,
                 ": terminating on i/o error reading stdin");
   } else {
      cstr_vcopy(attention, progname,
                 ": terminating normally on eof reading stdin");
   }
   logline_post(attention, 0);
   closelog();

   return ((e != 0) ? 111 : 0);
}


/* eof: sissylog.c */
